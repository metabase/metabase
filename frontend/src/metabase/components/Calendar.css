.Calendar-week {
  display: flex;
}

.Calendar-day,
.Calendar-day-name {
  flex: 1;
}

.Calendar-day {
  color: color(var(--base-grey) shade(30%));
  position: relative;
  border: 1px solid color(var(--base-grey) shade(20%) alpha(-50%));
  border-radius: 0;
  border-bottom-width: 0;
  border-right-width: 0;
}

.Calendar-day:last-child {
  border-right-width: 1px;
}
.Calendar-week:last-child .Calendar-day {
  border-bottom-width: 1px;
}

.Calendar-day-name {
  cursor: inherit;
}

.Calendar-day--this-month {
  color: currentcolor;
}

.Calendar-day--today {
  font-weight: 700;
}

.Calendar-day:hover {
  color: var(--purple-color);
}

.Calendar-day-name {
  color: inherit !important;
}

.Calendar-day--selected,
.Calendar-day--selected-end {
  color: white !important;
  background-color: var(--purple-color);
  z-index: 1;
}

.Calendar-day--in-range {
  background-color: #e3daeb;
}

.Calendar-day--selected:after,
.Calendar-day--selected-end:after,
.Calendar-day--in-range:after {
  content: "";
  position: absolute;
  top: -2px;
  bottom: -1px;
  left: -2px;
  right: -2px;
  border: 2px solid color(var(--purple-color) shade(25%));
  border-radius: 4px;
  z-index: 2;
}

.Calendar-day--in-range:after {
  border-left-color: transparent;
  border-right-color: transparent;
  border-radius: 0px;
}

.Calendar-day--week-start.Calendar-day--in-range:after {
  border-top-left-radius: 4px;
  border-bottom-left-radius: 4px;
  border-left-color: color(var(--purple-color) shade(25%));
}

.Calendar-day--week-end.Calendar-day--in-range:after {
  border-top-right-radius: 4px;
  border-bottom-right-radius: 4px;
  border-right-color: color(var(--purple-color) shade(25%));
}

.circle-button {
  display: block;
  font-size: 20px;
  color: color(var(--base-grey) shade(30%));
  border: 2px solid color(var(--base-grey) shade(10%));
  border-radius: 99px;
  width: 24px;
  height: 24px;
  background-color: white;
  text-align: center;
  vertical-align: middle;
  line-height: 20px;
  z-index: 2;
}

.circle-button:hover {
  color: var(--purple-color);
  border-color: var(--purple-color);
}

.circle-button--top {
  position: absolute;
  top: -12px;
}
.circle-button--bottom {
  position: absolute;
  bottom: -12px;
}
.circle-button--left {
  position: absolute;
  left: -12px;
}
.circle-button--right {
  position: absolute;
  right: -12px;
}

.Calendar--noContext .Calendar-day {
  visibility: hidden;
  pointer-events: none;
}

.Calendar--noContext .Calendar-day--this-month {
  visibility: visible;
  pointer-events: all;
}
