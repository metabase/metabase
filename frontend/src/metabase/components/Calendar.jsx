import React, { Component } from "react";
import PropTypes from "prop-types";

import "./Calendar.css";

import cx from "classnames";
import moment from "moment";
import { t } from "c-3po";
import Icon from "metabase/components/Icon";

export default class Calendar extends Component {
  constructor(props) {
    super(props);

    this.state = {
      current: moment(props.initial || undefined),
    };
  }

  static propTypes = {
    selected: PropTypes.object,
    selectedEnd: PropTypes.object,
    onChange: PropTypes.func.isRequired,
    isRangePicker: PropTypes.bool,
    isDual: PropTypes.bool,
  };

  static defaultProps = {
    isRangePicker: true,
  };

  componentWillReceiveProps(nextProps) {
    if (
      // `selected` became null or not null
      (nextProps.selected == null) !== (this.props.selected == null) ||
      // `selectedEnd` became null or not null
      (nextProps.selectedEnd == null) !== (this.props.selectedEnd == null) ||
      // `selected` is not null and doesn't match previous `selected`
      (nextProps.selected != null &&
        !moment(nextProps.selected).isSame(this.props.selected, "day")) ||
      // `selectedEnd` is not null and doesn't match previous `selectedEnd`
      (nextProps.selectedEnd != null &&
        !moment(nextProps.selectedEnd).isSame(this.props.selectedEnd, "day"))
    ) {
      let resetCurrent = false;
      if (nextProps.selected != null && nextProps.selectedEnd != null) {
        // reset if `current` isn't between `selected` and `selectedEnd` month
        resetCurrent =
          nextProps.selected.isAfter(this.state.current, "month") &&
          nextProps.selectedEnd.isBefore(this.state.current, "month");
      } else if (nextProps.selected != null) {
        // reset if `current` isn't in `selected` month
        resetCurrent =
          nextProps.selected.isAfter(this.state.current, "month") ||
          nextProps.selected.isBefore(this.state.current, "month");
      }
      if (resetCurrent) {
        this.setState({ current: nextProps.selected });
      }
    }
  }

  onClickDay = date => {
    let { selected, selectedEnd, isRangePicker } = this.props;
    if (!isRangePicker || !selected || selectedEnd) {
      this.props.onChange(date.format("YYYY-MM-DD"), null);
    } else if (!selectedEnd) {
      if (date.isAfter(selected)) {
        this.props.onChange(
          selected.format("YYYY-MM-DD"),
          date.format("YYYY-MM-DD"),
        );
      } else {
        this.props.onChange(
          date.format("YYYY-MM-DD"),
          selected.format("YYYY-MM-DD"),
        );
      }
    }
  };

  previous = () => {
    this.setState({ current: moment(this.state.current).add(-1, "M") });
  };

  next = () => {
    this.setState({ current: moment(this.state.current).add(1, "M") });
  };

  renderMonthHeader(current, side) {
    return (
      <div className="Calendar-header flex align-center">
        {side !== "right" && (
          <div
            className="bordered rounded p1 cursor-pointer transition-border border-hover px1"
            onClick={this.previous}
          >
            <Icon name="chevronleft" size={10} />
          </div>
        )}
        <span className="flex-full" />
        <h4 className="cursor-pointer rounded p1">
          {current.format("MMMM YYYY")}
        </h4>
        <span className="flex-full" />
        {side !== "left" && (
          <div
            className="bordered border-hover rounded p1 transition-border cursor-pointer px1"
            onClick={this.next}
          >
            <Icon name="chevronright" size={10} />
          </div>
        )}
      </div>
    );
  }

  renderDayNames() {
    const names = [t`Su`, t`Mo`, t`Tu`, t`We`, t`Th`, t`Fr`, t`Sa`];
    return (
      <div className="Calendar-day-names Calendar-week py1">
        {names.map(name => (
          <span key={name} className="Calendar-day-name text-centered">
            {name}
          </span>
        ))}
      </div>
    );
  }

  renderWeeks(current) {
    let weeks = [],
      done = false,
      date = moment(current)
        .startOf("month")
        .day("Sunday"),
      monthIndex = date.month(),
      count = 0;

    while (!done) {
      weeks.push(
        <Week
          key={date.toString()}
          date={moment(date)}
          month={current}
          onClickDay={this.onClickDay}
          selected={this.props.selected}
          selectedEnd={this.props.selectedEnd}
        />,
      );
      date.add(1, "w");
      done = count++ > 2 && monthIndex !== date.month();
      monthIndex = date.month();
    }

    return <div className="Calendar-weeks relative">{weeks}</div>;
  }

  renderCalender(current, side) {
    return (
      <div
        className={cx("Calendar Grid-cell", {
          "Calendar--range": this.props.selected && this.props.selectedEnd,
        })}
      >
        {this.renderMonthHeader(current, side)}
        {this.renderDayNames(current)}
        {this.renderWeeks(current)}
      </div>
    );
  }

  render() {
    const { current } = this.state;
    if (this.props.isDual) {
      return (
        <div className="Grid Grid--1of2 Grid--gutters">
          {this.renderCalender(current, "left")}
          {this.renderCalender(moment(current).add(1, "month"), "right")}
        </div>
      );
    } else {
      return this.renderCalender(current);
    }
  }
}

class Week extends Component {
  static propTypes = {
    selected: PropTypes.object,
    selectedEnd: PropTypes.object,
    onClickDay: PropTypes.func.isRequired,
  };

  render() {
    let days = [];
    let { date, month, selected, selectedEnd } = this.props;

    for (let i = 0; i < 7; i++) {
      let classes = cx("Calendar-day p1 cursor-pointer text-centered", {
        "Calendar-day--today": date.isSame(new Date(), "day"),
        "Calendar-day--this-month": date.month() === month.month(),
        "Calendar-day--selected": selected && date.isSame(selected, "day"),
        "Calendar-day--selected-end":
          selectedEnd && date.isSame(selectedEnd, "day"),
        "Calendar-day--week-start": i === 0,
        "Calendar-day--week-end": i === 6,
        "Calendar-day--in-range":
          !(date.isSame(selected, "day") || date.isSame(selectedEnd, "day")) &&
          (date.isSame(selected, "day") ||
            date.isSame(selectedEnd, "day") ||
            (selectedEnd &&
              selectedEnd.isAfter(date, "day") &&
              date.isAfter(selected, "day"))),
      });
      days.push(
        <span
          key={date.toString()}
          className={classes}
          onClick={this.props.onClickDay.bind(null, date)}
        >
          {date.date()}
        </span>,
      );
      date = moment(date).add(1, "d");
    }

    return (
      <div className="Calendar-week" key={days[0].toString()}>
        {days}
      </div>
    );
  }
}
