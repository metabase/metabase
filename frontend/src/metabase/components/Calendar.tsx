/* eslint-disable react/prop-types */
import React, { Component } from "react";
import PropTypes from "prop-types";

import "./Calendar.css";

import cx from "classnames";
import moment, { Moment } from "moment";
import { t } from "ttag";
import Icon from "metabase/components/Icon";
import { alpha, color } from "metabase/lib/colors";

export type SelectAll = "after" | "before";

type Props = {
  initial?: Moment;
  selected: Moment;
  selectedEnd?: Moment;
  selectAll?: SelectAll | null;
  onChange: (
    start: string,
    end: string | null,
    startMoment: Moment,
    endMoment: Moment | null,
  ) => void;
  isRangePicker?: boolean;
  primaryColor?: string;
};

type State = {
  current: Moment;
};

export default class Calendar extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      current: moment(props.initial || undefined),
    };
  }

  static propTypes = {};

  static defaultProps = {
    isRangePicker: true,
  };

  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    if (
      // `selected` became null or not null
      (nextProps.selected == null) !== (this.props.selected == null) ||
      // `selectedEnd` became null or not null
      (nextProps.selectedEnd == null) !== (this.props.selectedEnd == null) ||
      // `selected` is not null and doesn't match previous `selected`
      (nextProps.selected != null &&
        !moment(nextProps.selected).isSame(this.props.selected, "day")) ||
      // `selectedEnd` is not null and doesn't match previous `selectedEnd`
      (nextProps.selectedEnd != null &&
        !moment(nextProps.selectedEnd).isSame(this.props.selectedEnd, "day"))
    ) {
      let resetCurrent = false;
      if (nextProps.selected != null && nextProps.selectedEnd != null) {
        // reset if `current` isn't between `selected` and `selectedEnd` month
        resetCurrent =
          nextProps.selected.isAfter(this.state.current, "month") &&
          nextProps.selectedEnd.isBefore(this.state.current, "month");
      } else if (nextProps.selected != null) {
        // reset if `current` isn't in `selected` month
        resetCurrent =
          nextProps.selected.isAfter(this.state.current, "month") ||
          nextProps.selected.isBefore(this.state.current, "month");
      }
      if (resetCurrent) {
        this.setState({ current: nextProps.selected });
      }
    }
  }

  onClickDay = (date: Moment) => {
    const { selected, selectedEnd, isRangePicker } = this.props;
    if (!isRangePicker || !selected || selectedEnd) {
      this.props.onChange(date.format("YYYY-MM-DD"), null, date, null);
    } else if (!selectedEnd) {
      if (date.isAfter(selected)) {
        this.props.onChange(
          selected.format("YYYY-MM-DD"),
          date.format("YYYY-MM-DD"),
          selected,
          date,
        );
      } else {
        this.props.onChange(
          date.format("YYYY-MM-DD"),
          selected.format("YYYY-MM-DD"),
          date,
          selected,
        );
      }
    }
  };

  previous = () => {
    this.setState({ current: moment(this.state.current).add(-1, "M") });
  };

  next = () => {
    this.setState({ current: moment(this.state.current).add(1, "M") });
  };

  renderMonthHeader(current: Moment, side?: "left" | "right") {
    return (
      <div className="Calendar-header flex align-center border-bottom">
        {side !== "right" && (
          <div
            className="cursor-pointer text-brand-hover"
            onClick={this.previous}
          >
            <Icon name="chevronleft" size={10} />
          </div>
        )}
        <span className="flex-full" />
        <h4>{current.format("MMMM YYYY")}</h4>
        <span className="flex-full" />
        {side !== "left" && (
          <div className="cursor-pointer text-brand-hover" onClick={this.next}>
            <Icon name="chevronright" size={10} />
          </div>
        )}
      </div>
    );
  }

  renderDayNames() {
    const names = [t`Su`, t`Mo`, t`Tu`, t`We`, t`Th`, t`Fr`, t`Sa`];
    return (
      <div className="Calendar-day-names Calendar-week py1">
        {names.map(name => (
          <span key={name} className="Calendar-day-name text-centered">
            {name}
          </span>
        ))}
      </div>
    );
  }

  renderWeeks(current: Moment) {
    const weeks = [];
    const date = moment(current)
      .startOf("month")
      .day("Sunday");
    let done = false;
    let monthIndex = date.month();
    let count = 0;

    while (!done) {
      weeks.push(
        <Week
          key={date.toString()}
          date={moment(date)}
          month={current}
          onClickDay={this.onClickDay}
          isRangePicker={this.props.isRangePicker}
          selected={this.props.selected}
          selectedEnd={this.props.selectedEnd}
          selectAll={this.props.selectAll}
        />,
      );
      date.add(1, "w");
      done = count++ > 2 && monthIndex !== date.month();
      monthIndex = date.month();
    }

    return <div className="Calendar-weeks relative">{weeks}</div>;
  }

  renderCalender(current: Moment, side?: "left" | "right") {
    return (
      <div
        className={cx("Calendar", {
          "Calendar--range":
            (this.props.isRangePicker &&
              this.props.selected &&
              this.props.selectedEnd) ||
            this.props.selectAll,
        })}
      >
        {this.renderMonthHeader(current, side)}
        {this.renderDayNames()}
        {this.renderWeeks(current)}
      </div>
    );
  }

  render() {
    const { current } = this.state;
    return this.renderCalender(current);
  }
}

type WeekProps = {
  date: Moment;
  month: Moment;
  selected: Moment;
  selectedEnd?: Moment;
  selectAll?: SelectAll | null;
  isRangePicker?: boolean;
  primaryColor?: string;
  onClickDay: (date: Moment) => void;
};

class Week extends Component<WeekProps> {
  render() {
    const days = [];
    let {
      date,
      month,
      selected,
      selectedEnd,
      primaryColor = color("brand"),
      selectAll,
      isRangePicker,
    } = this.props;

    for (let i = 0; i < 7; i++) {
      const isSelected =
        date.isSame(selected, "day") ||
        (isRangePicker && date.isSame(selectedEnd, "day"));
      let inRange = false;
      if (
        selected &&
        date.isAfter(selected, "day") &&
        selectedEnd &&
        selectedEnd.isAfter(date, "day")
      ) {
        inRange = true;
      } else if (selectAll === "after") {
        inRange = !!(selected && date.isAfter(selected, "day"));
      } else if (selectAll === "before") {
        inRange = !!(selected && selected.isAfter(date, "day"));
      }
      const bgColor = isSelected ? primaryColor : alpha(primaryColor, 0.1);
      const isEnd = selectAll === "before" && date.isSame(selected, "day");
      const classes = cx("Calendar-day cursor-pointer text-centered", {
        "Calendar-day--today": date.isSame(new Date(), "day"),
        "Calendar-day--this-month": date.month() === month.month(),
        "Calendar-day--selected":
          !isEnd && selected && date.isSame(selected, "day"),
        "Calendar-day--selected-end":
          isEnd || (selectedEnd && date.isSame(selectedEnd, "day")),
        "Calendar-day--week-start": i === 0,
        "Calendar-day--week-end": i === 6,
        "Calendar-day--in-range":
          (selectAll && inRange) ||
          (!(date.isSame(selected, "day") || date.isSame(selectedEnd, "day")) &&
            (date.isSame(selected, "day") ||
              date.isSame(selectedEnd, "day") ||
              (selectedEnd &&
                selectedEnd.isAfter(date, "day") &&
                date.isAfter(selected, "day")))),
      });
      days.push(
        <span
          key={date.toString()}
          className={classes}
          onClick={this.props.onClickDay.bind(null, date)}
          style={{
            backgroundColor: isSelected || inRange ? bgColor : undefined,
            color: !isSelected && inRange ? primaryColor : undefined,
          }}
        >
          {date.date()}
        </span>,
      );
      date = moment(date).add(1, "d");
    }

    return (
      <div className="Calendar-week" key={days[0].toString()}>
        {days}
      </div>
    );
  }
}
