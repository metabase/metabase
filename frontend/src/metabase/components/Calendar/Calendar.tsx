import React, { Component } from "react";
import cx from "classnames";
import moment, { Moment } from "moment-timezone";
import Icon from "metabase/components/Icon";
import {
  getDayOfWeekOptions,
  getFirstDayOfWeekIndex,
} from "metabase/lib/date-time";

import "./Calendar.css";
import { CalendarDay } from "./Calendar.styled";

export type SelectAll = "after" | "before";

export type CalendarProps = {
  initial?: Moment;
  selected?: Moment;
  selectedEnd?: Moment;
  selectAll?: SelectAll | null;
  onChange?: (
    start: string,
    end: string | null,
    startMoment: Moment,
    endMoment?: Moment | null,
  ) => void;
  onChangeDate?: (date: string, dateMoment: Moment) => void;
  isRangePicker?: boolean;
  primaryColor?: string;
};

type State = {
  current?: Moment;
};

export default class Calendar extends Component<CalendarProps, State> {
  constructor(props: CalendarProps) {
    super(props);
    this.state = {
      current: moment(props.initial || undefined),
    };
  }

  static defaultProps = {
    isRangePicker: true,
  };

  UNSAFE_componentWillReceiveProps(nextProps: CalendarProps) {
    if (
      // `selected` became null or not null
      (nextProps.selected == null) !== (this.props.selected == null) ||
      // `selectedEnd` became null or not null
      (nextProps.selectedEnd == null) !== (this.props.selectedEnd == null) ||
      // `selected` is not null and doesn't match previous `selected`
      (nextProps.selected != null &&
        !moment(nextProps.selected).isSame(this.props.selected, "day")) ||
      // `selectedEnd` is not null and doesn't match previous `selectedEnd`
      (nextProps.selectedEnd != null &&
        !moment(nextProps.selectedEnd).isSame(this.props.selectedEnd, "day"))
    ) {
      let resetCurrent = false;
      if (nextProps.selected != null && nextProps.selectedEnd != null) {
        // reset if `current` isn't between `selected` and `selectedEnd` month
        resetCurrent =
          nextProps.selected.isAfter(this.state.current, "month") &&
          nextProps.selectedEnd.isBefore(this.state.current, "month");
      } else if (nextProps.selected != null) {
        // reset if `current` isn't in `selected` month
        resetCurrent =
          nextProps.selected.isAfter(this.state.current, "month") ||
          nextProps.selected.isBefore(this.state.current, "month");
      }
      if (resetCurrent) {
        this.setState({ current: nextProps.selected });
      }
    }
  }

  onClickDay = (date: Moment) => {
    const { selected, selectedEnd, isRangePicker } = this.props;

    if (!isRangePicker || !selected || selectedEnd) {
      this.props.onChange?.(date.format("YYYY-MM-DD"), null, date, null);
    } else if (!selectedEnd) {
      if (date.isAfter(selected)) {
        this.props.onChange?.(
          selected.format("YYYY-MM-DD"),
          date.format("YYYY-MM-DD"),
          selected,
          date,
        );
      } else {
        this.props.onChange?.(
          date.format("YYYY-MM-DD"),
          selected.format("YYYY-MM-DD"),
          date,
          selected,
        );
      }
    }

    this.props.onChangeDate?.(date.format("YYYY-MM-DD"), date);
  };

  previous = () => {
    this.setState({ current: moment(this.state.current).add(-1, "M") });
  };

  next = () => {
    this.setState({ current: moment(this.state.current).add(1, "M") });
  };

  renderMonthHeader(current?: Moment, side?: "left" | "right") {
    current = current || moment();
    return (
      <div className="Calendar-header flex align-center border-bottom">
        {side !== "right" && (
          <div
            className="cursor-pointer text-brand-hover"
            onClick={this.previous}
          >
            <Icon name="chevronleft" size={10} />
          </div>
        )}
        <span className="flex-full" />
        <h4>{current.format("MMMM YYYY")}</h4>
        <span className="flex-full" />
        {side !== "left" && (
          <div className="cursor-pointer text-brand-hover" onClick={this.next}>
            <Icon name="chevronright" size={10} />
          </div>
        )}
      </div>
    );
  }

  renderDayNames() {
    const days = getDayOfWeekOptions();

    return (
      <div className="Calendar-day-names Calendar-week py1">
        {days.map(({ shortName }) => (
          <span
            key={shortName}
            className="Calendar-day-name text-centered"
            data-testid="calendar-day-name"
          >
            {shortName}
          </span>
        ))}
      </div>
    );
  }

  renderWeeks(current?: Moment) {
    current = current || moment();
    const weeks = [];
    const firstDayOfWeek = getFirstDayOfWeekIndex();
    const date = moment(current).startOf("month").isoWeekday(firstDayOfWeek);

    // if set week doesn't start with 1st day of month, then add the previous week
    if (date.date() > 1) {
      date.add(-1, "w");
    }

    let done = false;
    let monthIndex = date.month();
    let count = 0;

    while (!done) {
      weeks.push(
        <Week
          key={date.toString()}
          date={moment(date)}
          month={current}
          onClickDay={this.onClickDay}
          isRangePicker={this.props.isRangePicker}
          selected={this.props.selected}
          primaryColor={this.props.primaryColor}
          selectedEnd={this.props.selectedEnd}
          selectAll={this.props.selectAll}
        />,
      );
      date.add(1, "w");
      done = count++ > 2 && monthIndex !== date.month();
      monthIndex = date.month();
    }

    return (
      <div className="Calendar-weeks relative" data-testid="calendar-weeks">
        {weeks}
      </div>
    );
  }

  renderCalender(current?: Moment, side?: "left" | "right") {
    return (
      <div
        data-testid="calendar"
        className={cx("Calendar", {
          "Calendar--range":
            (this.props.isRangePicker &&
              this.props.selected &&
              this.props.selectedEnd) ||
            this.props.selectAll,
        })}
      >
        {this.renderMonthHeader(current, side)}
        {this.renderDayNames()}
        {this.renderWeeks(current)}
      </div>
    );
  }

  render() {
    const { current } = this.state;
    return this.renderCalender(current);
  }
}

type WeekProps = {
  date: Moment;
  month: Moment;
  selected?: Moment;
  selectedEnd?: Moment;
  selectAll?: SelectAll | null;
  isRangePicker?: boolean;
  primaryColor?: string;
  onClickDay: (date: Moment) => void;
};

class Week extends Component<WeekProps> {
  render() {
    const days = [];
    let { date, month, selected, selectedEnd, selectAll, isRangePicker } =
      this.props;

    for (let i = 0; i < 7; i++) {
      const isSelected =
        date.isSame(selected, "day") ||
        (isRangePicker &&
          selectedEnd?.isAfter(selected) &&
          date.isSame(selectedEnd, "day"));
      let isInRange = false;
      if (
        selected &&
        date.isAfter(selected, "day") &&
        selectedEnd &&
        selectedEnd.isAfter(date, "day")
      ) {
        isInRange = true;
      } else if (selectAll === "after") {
        isInRange = !!(selected && date.isAfter(selected, "day"));
      } else if (selectAll === "before") {
        isInRange = !!(selected && selected.isAfter(date, "day"));
      }
      const isEnd = selectAll === "before" && date.isSame(selected, "day");
      const isSelectedStart =
        !isEnd && selected && date.isSame(selected, "day");
      const isSelectedEnd =
        isEnd || (selectedEnd && date.isSame(selectedEnd, "day"));
      const classes = cx("Calendar-day cursor-pointer text-centered", {
        "Calendar-day--this-month": date.month() === month.month(),
        "Calendar-day--selected": isSelectedStart,
        "Calendar-day--selected-end": isSelectedEnd,
        "Calendar-day--week-start": i === 0,
        "Calendar-day--week-end": i === 6,
        "Calendar-day--in-range":
          (selectAll && isInRange) ||
          (!(date.isSame(selected, "day") || date.isSame(selectedEnd, "day")) &&
            (date.isSame(selected, "day") ||
              date.isSame(selectedEnd, "day") ||
              (selectedEnd &&
                selectedEnd.isAfter(date, "day") &&
                date.isAfter(selected, "day")))),
      });
      days.push(
        <CalendarDay
          key={date.toString()}
          className={classes}
          onClick={this.props.onClickDay.bind(null, date)}
          isInRange={isInRange}
          isSelected={isSelected}
          isSelectedStart={isSelectedStart}
          isSelectedEnd={isSelectedEnd}
          primaryColor={this.props.primaryColor}
        >
          {date.date()}
        </CalendarDay>,
      );
      date = moment(date).add(1, "d");
    }

    return (
      <div className="Calendar-week" key={days[0].toString()}>
        {days}
      </div>
    );
  }
}
