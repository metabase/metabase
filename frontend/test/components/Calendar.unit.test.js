import React from "react";
import renderer from "react-test-renderer";
import moment from "moment";
import mockDate from "mockdate";
import { mount } from "enzyme";

import Calendar from "../../src/metabase/components/Calendar";

describe("Calendar", () => {
  afterEach(() => {
    mockDate.reset();
  });

  it("should render correctly", () => {
    // set the system clock to the snapshot's current date
    mockDate.set("2018-01-12T12:00:00Z", 0);
    const tree = renderer
      .create(<Calendar selected={moment("2018-01-01")} onChange={() => {}} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should switch months correctly", () => {
    mockDate.set("2018-01-12T12:00:00Z", 0);
    const calendar = mount(
      <Calendar selected={moment("2018-01-01")} onChange={() => {}} />,
    );
    expect(calendar.find(".Calendar-header").text()).toEqual("January 2018");
    calendar.find(".Icon-chevronleft").simulate("click");
    expect(calendar.find(".Calendar-header").text()).toEqual("December 2017");
    calendar.find(".Icon-chevronright").simulate("click");
    calendar.find(".Icon-chevronright").simulate("click");
    expect(calendar.find(".Calendar-header").text()).toEqual("February 2018");
  });
});
