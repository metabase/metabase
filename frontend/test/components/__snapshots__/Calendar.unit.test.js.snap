// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar should render correctly 1`] = `
<div
  className="Calendar Grid-cell"
>
  <div
    className="Calendar-header flex align-center"
  >
    <div
      className="bordered rounded p1 cursor-pointer transition-border border-hover px1"
      onClick={[Function]}
    >
      <svg
        className="Icon Icon-chevronleft Icon-cxuQhR kTAgZA"
        fill="currentcolor"
        height={10}
        name="chevronleft"
        size={10}
        viewBox="0 0 32 32"
        width={10}
      >
        <path
          d="M20 1 L24 5 L14 16 L24 27 L20 31 L6 16 z"
        />
      </svg>
    </div>
    <span
      className="flex-full"
    />
    <h4
      className="cursor-pointer rounded p1"
    >
      January 2018
    </h4>
    <span
      className="flex-full"
    />
    <div
      className="bordered border-hover rounded p1 transition-border cursor-pointer px1"
      onClick={[Function]}
    >
      <svg
        className="Icon Icon-chevronright Icon-cxuQhR kTAgZA"
        fill="currentcolor"
        height={10}
        name="chevronright"
        size={10}
        viewBox="0 0 32 32"
        width={10}
      >
        <path
          d="M12 1 L26 16 L12 31 L8 27 L18 16 L8 5 z "
        />
      </svg>
    </div>
  </div>
  <div
    className="Calendar-day-names Calendar-week py1"
  >
    <span
      className="Calendar-day-name text-centered"
    >
      Su
    </span>
    <span
      className="Calendar-day-name text-centered"
    >
      Mo
    </span>
    <span
      className="Calendar-day-name text-centered"
    >
      Tu
    </span>
    <span
      className="Calendar-day-name text-centered"
    >
      We
    </span>
    <span
      className="Calendar-day-name text-centered"
    >
      Th
    </span>
    <span
      className="Calendar-day-name text-centered"
    >
      Fr
    </span>
    <span
      className="Calendar-day-name text-centered"
    >
      Sa
    </span>
  </div>
  <div
    className="Calendar-weeks relative"
  >
    <div
      className="Calendar-week"
    >
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--week-start"
        onClick={[Function]}
      >
        31
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--selected"
        onClick={[Function]}
      >
        1
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        2
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        3
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        4
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        5
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--week-end"
        onClick={[Function]}
      >
        6
      </span>
    </div>
    <div
      className="Calendar-week"
    >
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--week-start"
        onClick={[Function]}
      >
        7
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        8
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        9
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        10
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        11
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--today Calendar-day--this-month"
        onClick={[Function]}
      >
        12
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--week-end"
        onClick={[Function]}
      >
        13
      </span>
    </div>
    <div
      className="Calendar-week"
    >
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--week-start"
        onClick={[Function]}
      >
        14
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        15
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        16
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        17
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        18
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        19
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--week-end"
        onClick={[Function]}
      >
        20
      </span>
    </div>
    <div
      className="Calendar-week"
    >
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--week-start"
        onClick={[Function]}
      >
        21
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        22
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        23
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        24
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        25
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        26
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--week-end"
        onClick={[Function]}
      >
        27
      </span>
    </div>
    <div
      className="Calendar-week"
    >
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month Calendar-day--week-start"
        onClick={[Function]}
      >
        28
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        29
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        30
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--this-month"
        onClick={[Function]}
      >
        31
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered"
        onClick={[Function]}
      >
        1
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered"
        onClick={[Function]}
      >
        2
      </span>
      <span
        className="Calendar-day p1 cursor-pointer text-centered Calendar-day--week-end"
        onClick={[Function]}
      >
        3
      </span>
    </div>
  </div>
</div>
`;
